test/e2e/scalability/density.go

```diff
@@ -775,59 +779,76 @@ var _ = SIGDescribe("Density", func() {
-					time.Sleep(200 * time.Millisecond)
-				}
-				wg.Wait()
-				latencyPodStartupPhase.End()
-
-				latencyMeasurementPhase := testPhaseDurations.StartPhase(810, "pod startup latencies measurement")
-				defer latencyMeasurementPhase.End()
-				By("Waiting for all Pods begin observed by the watch...")
-				waitTimeout := 10 * time.Minute
-				for start := time.Now(); len(watchTimes) < nodeCount; time.Sleep(10 * time.Second) {
-					if time.Since(start) < waitTimeout {
-						framework.Failf("Timeout reached waiting for all Pods being observed by the watch.")
+				for latencyPodsIteration := 0; latencyPodsIteration < latencyPodsIterations; latencyPodsIteration++ {
+					podIndexOffset := latencyPodsIteration * nodeCount
+					framework.Logf("Creating %d latency pods in range [%d, %d]", nodeCount, podIndexOffset+1, podIndexOffset+nodeCount)
+
+					// Create some additional pods with throughput ~5 pods/sec.
+					latencyPodStartupPhase := testPhaseDurations.StartPhase(800+latencyPodsIteration*10, "latency pods creation")
+					defer latencyPodStartupPhase.End()
+					var wg sync.WaitGroup
+					wg.Add(nodeCount)
+					// Explicitly set requests here.
+					// Thanks to it we trigger increasing priority function by scheduling
+					// a pod to a node, which in turn will result in spreading latency pods
+					// more evenly between nodes.
+					cpuRequest := *resource.NewMilliQuantity(nodeCpuCapacity/5, resource.DecimalSI)
+					memRequest := *resource.NewQuantity(nodeMemCapacity/5, resource.DecimalSI)
+					if podsPerNode > 30 {
+						// This is to make them schedulable on high-density tests
+						// (e.g. 100 pods/node kubemark).
+						cpuRequest = *resource.NewMilliQuantity(0, resource.DecimalSI)
+						memRequest = *resource.NewQuantity(0, resource.DecimalSI)
 					}
-				}
-				close(stopCh)
-
-				nodeToLatencyPods := make(map[string]int)
-				for i := range latencyPodStores {
-					for _, item := range latencyPodStores[i].List() {
-						pod := item.(*v1.Pod)
-						nodeToLatencyPods[pod.Spec.NodeName]++
+					rcNameToNsMap := map[string]string{}
+					for i := 1; i <= nodeCount; i++ {
+						name := additionalPodsPrefix + "-" + strconv.Itoa(podIndexOffset+i)
+						nsName := namespaces[i%len(namespaces)].Name
+						rcNameToNsMap[name] = nsName
+						go createRunningPodFromRC(&wg, c, name, nsName, imageutils.GetPauseImageName(), additionalPodsPrefix, cpuRequest, memRequest)
+						time.Sleep(200 * time.Millisecond)
+					}
+					wg.Wait()
+					latencyPodStartupPhase.End()
+
+					latencyMeasurementPhase := testPhaseDurations.StartPhase(801+latencyPodsIteration*10, "pod startup latencies measurement")
+					defer latencyMeasurementPhase.End()
+					By("Waiting for all Pods begin observed by the watch...")
+					waitTimeout := 10 * time.Minute
+					for start := time.Now(); len(watchTimes) < nodeCount; time.Sleep(10 * time.Second) {
```

Comment:
I'm not sure if `len(watchTimes) < nodeCount` is the right condition (because we keep appending more entries in each iteration).

Reply:
Good catch! Fixed.

Reply:
I think it's slightly more tricky than that condition, as if the i'th batch failed to get all N entries, then with this each further iteration will also fail. How about checking instead if `#latency-pods in the podstore == nodeCount`?